src/Renderer/GDLibRenderer.php

```diff
@@ -0,0 +1,248 @@
+     */
+    private function normalizePoints(array $points): array
+    {
+        $maxX = $maxY = 0;
+        for ($i = 0; $i < count($points); $i += 2) {
+            // Do manual round as GD just removes decimal part
+            $points[$i] = $newX = round($points[$i]);
+            $points[$i + 1] = $newY = round($points[$i + 1]);
+
+            $maxX = max($maxX, $newX);
+            $maxY = max($maxY, $newY);
+        }
+
+        // Do trimming only if there are 4 points (8 coordinates), assumes this is square.
+
+        for ($i = 0; $i < count($points); $i += 2) {
+            $points[$i] = min($points[$i], $maxX - 1);
+            $points[$i + 1] = min($points[$i + 1], $maxY - 1);
+        }
+
+        return $points;
+    }
+
+    private function renderImage(): string
+    {
+        ob_start();
+        $quality = $this->compressionQuality;
+        switch ($this->imageFormat) {
+            case 'png':
+                if ($quality > 9 || $quality < 0) {
+                    $quality = 9;
+                }
+                imagepng($this->image, null, $quality);
+                break;
+
+            case 'gif':
+                imagegif($this->image, null);
+                break;
+
+            case 'jpeg':
+            case 'jpg':
+                if ($quality > 100 || $quality < 0) {
+                    $quality = 85;
+                }
+                imagejpeg($this->image, null, $quality);
+                break;
+            default:
+                ob_end_clean();
+                throw new InvalidArgumentException(
+                    'Supported image formats are jpeg, png and gif, got: ' . $this->imageFormat
+                );
+        }
+
+        imagedestroy($this->image);
+        $this->colors = [];
+        $this->image = null;
+
+        return ob_get_clean();
+    }
+
+    private function getColor(ColorInterface $color): int
+    {
+        $alpha = 100;
+
+        if ($color instanceof Alpha) {
+            $alpha = $color->getAlpha();
+            $color = $color->getBaseColor();
+        }
+
+        $rgb = $color->toRgb();
+
+        $colorKey = sprintf('%02X%02X%02X%02X', $rgb->getRed(), $rgb->getGreen(), $rgb->getBlue(), $alpha);
+
+        if (! isset($this->colors[$colorKey])) {
+            $colorId = imagecolorallocatealpha(
+                $this->image,
+                $rgb->getRed(),
+                $rgb->getGreen(),
+                $rgb->getBlue(),
+                (int)((100 - $alpha) / 100 * 127) // Alpha for GD is in range 0 (opaque) - 127 (transparent)
+            );
+
+            if ($colorId === false) {
+                throw new RuntimeException('Failed to create color: #' . $colorKey);
+            }
+
+            $this->colors[$colorKey] = $colorId;
+        }
+
+        return $this->colors[$colorKey];
+    }
+
+    private function imageFilledPolygon(array $points, int $color): bool
+    {
+        if (\PHP_VERSION_ID >= 80000) {
```

Comment:
This check should not be required anymore, as we raised the minimum version to 8.1 in the next release.

Reply:
Removed